### ISSUE TITLE
Verilator Unexpected Crash at traceAll

### ISSUE THREAD

**Issue by DenizzzGuzell** [verilator/verilator#5372]
Hello 

A project that uses an embedded FPGA with adjustable tile size has been under my verification. There were no issues when verifying low tile sizes. But when tile sizes grow, designs become more intricate, and Verilator crashes as shown in the part below: 

```
- V3GraphAlg.cpp:462: Order:
- V3OrderProcessDomains.cpp:82:  Domains...
- V3OrderSerial.cpp:37:  Constructing serial code for 'stl'- V3SchedPartition.cpp:334:partition: 
- V3GraphAlg.cpp:462: Order:
- V3OrderProcessDomains.cpp:82:  Domains...
- V3OrderSerial.cpp:37:  Constructing serial code for 'ico'- V3GraphAlg.cpp:462: Order:
- V3OrderProcessDomains.cpp:82:  Domains...
- V3OrderSerial.cpp:37:  Constructing serial code for 'act'- V3GraphAlg.cpp:462: Order:
- V3OrderProcessDomains.cpp:82:  Domains...
- V3OrderSerial.cpp:37:  Constructing serial code for 'nba'- V3Clock.cpp:229:    clockAll: 
- V3Const.cpp:3988:   constifyAll: 
- V3Life.cpp:497:     lifeAll: 
- V3LifePost.cpp:362: lifepostAll: 
- V3Const.cpp:3988:   constifyAll: 
- V3Dead.cpp:568:     deadifyAllScoped: 
- V3Trace.cpp:913:    traceAll: 
%Error: export VERILATOR_ROOT=/home/dguzel/verilator
%Error: ulimit -s unlimited 2>/dev/null; exec /home/dguzel/verilator/bin/verilator_bin_dbg -cc -CFLAGS -std=c++14 --exe --x-assign 0 -Wno-lint --trace --assert --build -j --debug --no-debug-leak -no-dump-tree --top-module fpga_top --autoflush -F ../Fabric/SRC/fabric_netlists.flist verilator-main.cpp
%Error: Command Failed ulimit -s unlimited 2>/dev/null; exec /home/dguzel/verilator/bin/verilator_bin_dbg -cc -CFLAGS -std=c++14 --exe --x-assign 0 -Wno-lint --trace --assert --build -j --debug --no-debug-leak -no-dump-tree --top-module fpga_top --autoflush -F ../Fabric/SRC/fabric_netlists.flist verilator-main.cpp
Makefile:63: recipe for target 'verilate' failed
```

I have pasted here last part of the log file which is generated debug flag on.

During the Verilation process, Verilator consumes approximately 200GB of system memory—about 90% of the total available resources—and utilizes 100% of the CPU. This intensive resource consumption continues for nearly 90 minutes before the process ultimately crashes.

I have enabled the debug flag, and while the log file was generated, it does not provide any specific error messages other than three unspecified ones at the end of the process. Additionally, I have already set the terminal ulimit to unlimited, so I believe that the crash might be due to the unsufficent source.

Could you please clarify whether this crash is indeed a result of the design's complexity? If not, could you help me understand the underlying cause and suggest ways to optimize the process to prevent Verilator from crashing?

Thank you for your assistance.

**Comment by DenizzzGuzell** [comment]
Before V3Trace stage memory usage is around 60 gb when traceAll starts it increases quite fast to 200 210 gb (maybe more I can not check usage because OS stucks)then crashes. What you should consider is coming to the stage where traceAll start is takes more than an hour. 

Model builded ok without --trace signal below you can find stats text file. This process also took a quite long time.
[Vfpga_top__stats.txt](https://github.com/user-attachments/files/16622532/Vfpga_top__stats.txt)

**Comment by wsnyder** [comment]
Do these seem roughly right?
```
Input, Verilog bytes read                              187,077,522
Output, C++ bytes written                        3,400,715,433
Node count, VARSCOPE                                     747,823
```

It's going to make a structure for each varscope, but at 200 GB/747k but needing 256K/signal seems broken.

You could try running under GDB, hit ctrl-c near where memory is getting huge, and seeing where it is.  Alternatively start adding prints in V3Trace to see which "new" is using most of the space.

**Comment by DenizzzGuzell** [comment]
Hello,

I have configured a GDB environment to monitor memory allocation operations and system call memory accesses. Based on my observations, the memory consumption reaches approximately 100GB within 5 minutes, specifically within the methods identified in the backtrace provided below:

```
Breakpoint 7, 0x0000555556027fb0 in sbrk ()
(gdb) bt
#0  0x0000555556027fb0 in sbrk ()
#1  0x0000555555fc3029 in SbrkSysAllocator::Alloc(unsigned long, unsigned long*, unsigned long) ()
#2  0x0000555555fc33d6 in DefaultSysAllocator::Alloc(unsigned long, unsigned long*, unsigned long) ()
#3  0x0000555555fc3546 in TCMalloc_SystemAlloc(unsigned long, unsigned long*, unsigned long) ()
#4  0x0000555555fc1a06 in tcmalloc::PageHeap::GrowHeap(unsigned long) ()
#5  0x0000555555fc1d4b in tcmalloc::PageHeap::New(unsigned long) ()
#6  0x0000555555fbc30f in tcmalloc::CentralFreeList::Populate() ()
#7  0x0000555555fbc518 in tcmalloc::CentralFreeList::FetchFromOneSpansSafe(int, void**, void**) ()
#8  0x0000555555fbc60f in tcmalloc::CentralFreeList::RemoveRange(void**, void**, int) ()
#9  0x0000555555fc389a in tcmalloc::ThreadCache::FetchFromCentralCache(unsigned long, unsigned long) ()
#10 0x000055555602780d in tc_newarray ()
#11 0x0000555555f03d66 in TraceVisitor::addActivitySetter (this=this@entry=0x7fffffffdc20, insertp=0x55574befce60, code=code@entry=12508) at ../V3Trace.cpp:401
#12 0x0000555555f04640 in TraceVisitor::createActivityFlags (this=this@entry=0x7fffffffdc20) at ../V3Trace.cpp:448
#13 0x0000555555f08f45 in TraceVisitor::createTraceFunctions (this=this@entry=0x7fffffffdc20) at ../V3Trace.cpp:734
#14 0x0000555555f094cb in TraceVisitor::visit (this=0x7fffffffdc20, nodep=0x555556e2c100) at ../V3Trace.cpp:810
#15 0x00005555559b05f2 in AstNetlist::accept (this=<optimized out>, v=...) at ../V3AstNodeOther.h:1302
#16 0x0000555555efe6e2 in VNVisitor::iterate (nodep=0x555556e2c100, this=0x7fffffffdc20) at ../V3Ast.h:2936
#17 TraceVisitor::TraceVisitor (nodep=0x555556e2c100, this=0x7fffffffdc20) at ../V3Trace.cpp:901
#18 V3Trace::traceAll (nodep=0x555556e2c100) at ../V3Trace.cpp:914
#19 0x000055555595b6cf in process () at ../V3Global.h:143
#20 0x000055555595c664 in verilate (argString=...) at ../Verilator.cpp:707
#21 0x000055555595e531 in main (argc=<optimized out>, argv=0x7fffffffe028) at ../Verilator.cpp:852
```
The "rdi" size, which represents the memory allocation size, has been observed to be 1MB. If it seems improbable that this method is responsible for consuming such a significant amount of memory, it may indicate an incorrect memory operation or a potential misunderstanding on my part. Therefore, I intend to provide a more detailed and complex log file, which includes all the commands executed using GDB, for a thorough examination and clearer understanding.

[Memory_Overflow_Verilator.log](https://github.com/user-attachments/files/16637437/Memory_Overflow_Verilator.log)


Thank you.

**Comment by wsnyder** [comment]
Note we are awaiting a test case, or ideally a pull request to fix the issue, before we can proceed further, thanks.

**Comment by wsnyder** [comment]
It looks like V3SchedVirtInterface is adding a crazy number of assignments - it's also by far the slowest processing stage up until that point.  Then V3Tracing makes a vertex to track all of that information, which is not ideal, but not really the root issue.

@kbieganski can you please look into what we can do to improve V3SchedVirtInterface for this, thanks.

**Comment by wsnyder** [comment]
BTW to debug I've been using `--debugi 2 --no-dump-tree --no-debug-leak`

Then hacked in to dump only before the stage of interest:

```
diff --git a/src/V3SchedVirtIface.cpp b/src/V3SchedVirtIface.cpp
--- a/src/V3SchedVirtIface.cpp
+++ b/src/V3SchedVirtIface.cpp
@@ -224,6 +224,7 @@ public:
 VirtIfaceTriggers makeVirtIfaceTriggers(AstNetlist* nodep) {
     UINFO(2, __FUNCTION__ << ": " << endl);
     if (v3Global.hasVirtIfaces()) {
+        V3Global::dumpCheckGlobalTree("pre_sched_vif", 0, dumpTreeEitherLevel() >= 1);
         VirtIfaceVisitor visitor{nodep};
         V3Global::dumpCheckGlobalTree("sched_vif", 0, dumpTreeEitherLevel() >= 6);
         return visitor.take_triggers();
```

**Comment by wsnyder** [comment]
@kbieganski, hold off for now, digging in a bit more...

**Comment by DenizzzGuzell** [comment]
Hi @wsnyder ,

It looks like some of the comments disappeared. Do you expect something further from me related with given issue ?

**Comment by wsnyder** [comment]
I got it to run but I have only 32G so it runs out of memory before the trace stage.  Is there an easy way to make it e.g. 4-10x smaller?

Note as a workaround use the trace control metacomments to only trace the top level signals, what's causing problems is tracking all of the low level muxes and flops.

Also if you'd like to work on fixing it that would always be appreciated.

**Comment by DenizzzGuzell** [comment]
Hi @wsnyder, 

I have prepared an environment which uses 10x10 tile size and makes things less complicated. My test results generally shows approximately 10GB of system memory usage with using `-trace` flag without metacomments. You can verilate it under eFPGA_10x10/sim given repository link below.

For utilization, I have discovered a flag which traces only modules which are starts with `_` symbol as `--trace-underscore`. So I integrated this flag to 20x20 side and change top module name from ``fpga_top`` to `` _fpga_top``. With this approximation, it currently consumes approximately 42GB system memory to verilate. I also kept this size in repository if further investigation needed. 

https://github.com/DenizzzGuzell/eFPGA

Thank you

**Comment by wsnyder** [comment]
That was helpful, due to the odd scheduling it went almost O(n^2).

Prev V3Trace time 57 seconds, new time 12 seconds. Memory shrunk by 14 GB.

**Comment by wsnyder** [comment]
Please confirm it works on your larger model, thanks.

**Comment by DenizzzGuzell** [comment]
Hi @wsnyder,

I cannot build main branch due to given error below: 

`include V3PchAstNoMT.h -c ../V3Randomize.cpp -o V3Randomize.o
In file included from ../V3PchAstNoMT.h:29:
../V3Randomize.cpp: In static member function ‘static AstNodeStmt* RandomizeVisitor::wrapIfRandMode(AstClass*, AstVar*, AstNodeStmt*)’:
../V3Randomize.cpp:1210:88: error: no matching function for call to ‘RandomizeVisitor::wrapIfMode(<brace-enclosed initializer list>, AstVar*, AstNodeStmt*&)’
         return VN_AS(wrapIfMode({.asInt = varp->user1()}, getRandModeVar(classp), stmtp),
                                                                                        ^
../V3Ast.h:81:84: note: in definition of macro ‘VN_AS’
 #define VN_AS(nodep, nodetypename) (AstNode::privateAs<Ast##nodetypename, decltype(nodep)>(nodep))
                                                                                    ^~~~~
../V3Randomize.cpp:1217:21: note: candidate: ‘static AstNode* RandomizeVisitor::wrapIfMode(RandomizeMode, AstVar*, AstNode*)’
     static AstNode* wrapIfMode(const RandomizeMode mode, AstVar* modeVarp, AstNode* stmtp) {
                     ^~~~~~~~~~
../V3Randomize.cpp:1217:21: note:   no known conversion for argument 1 from ‘<brace-enclosed initializer list>’ to ‘RandomizeMode’
In file included from ../V3PchAstNoMT.h:29:
../V3Randomize.cpp:1210:88: error: no matching function for call to ‘RandomizeVisitor::wrapIfMode(<brace-enclosed initializer list>, AstVar*, AstNodeStmt*&)’
         return VN_AS(wrapIfMode({.asInt = varp->user1()}, getRandModeVar(classp), stmtp),
                                                                                        ^
../V3Ast.h:81:84: note: in definition of macro ‘VN_AS’
 #define VN_AS(nodep, nodetypename) (AstNode::privateAs<Ast##nodetypename, decltype(nodep)>(nodep))
                                                                                    ^~~~~
../V3Randomize.cpp:1217:21: note: candidate: ‘static AstNode* RandomizeVisitor::wrapIfMode(RandomizeMode, AstVar*, AstNode*)’
     static AstNode* wrapIfMode(const RandomizeMode mode, AstVar* modeVarp, AstNode* stmtp) {
                     ^~~~~~~~~~
../V3Randomize.cpp:1217:21: note:   no known conversion for argument 1 from ‘<brace-enclosed initializer list>’ to ‘RandomizeMode’
In file included from ../V3PchAstNoMT.h:29:
../V3Randomize.cpp:1210:88: error: no matching function for call to ‘RandomizeVisitor::wrapIfMode(<brace-enclosed initializer list>, AstVar*, AstNodeStmt*&)’
         return VN_AS(wrapIfMode({.asInt = varp->user1()}, getRandModeVar(classp), stmtp),
                                                                                        ^
../V3Ast.h:81:92: note: in definition of macro ‘VN_AS’
 #define VN_AS(nodep, nodetypename) (AstNode::privateAs<Ast##nodetypename, decltype(nodep)>(nodep))
                                                                                            ^~~~~
../V3Randomize.cpp:1217:21: note: candidate: ‘static AstNode* RandomizeVisitor::wrapIfMode(RandomizeMode, AstVar*, AstNode*)’
     static AstNode* wrapIfMode(const RandomizeMode mode, AstVar* modeVarp, AstNode* stmtp) {
                     ^~~~~~~~~~
../V3Randomize.cpp:1217:21: note:   no known conversion for argument 1 from ‘<brace-enclosed initializer list>’ to ‘RandomizeMode’
../V3Randomize.cpp: In static member function ‘static AstNode* RandomizeVisitor::wrapIfConstraintMode(AstClass*, AstConstraint*, AstNode*)’:
../V3Randomize.cpp:1215:91: error: no matching function for call to ‘RandomizeVisitor::wrapIfMode(<brace-enclosed initializer list>, AstVar*, AstNode*&)’
         return wrapIfMode({.asInt = constrp->user1()}, getConstraintModeVar(classp), stmtp);
                                                                                           ^
../V3Randomize.cpp:1217:21: note: candidate: ‘static AstNode* RandomizeVisitor::wrapIfMode(RandomizeMode, AstVar*, AstNode*)’
     static AstNode* wrapIfMode(const RandomizeMode mode, AstVar* modeVarp, AstNode* stmtp) {
                     ^~~~~~~~~~
../V3Randomize.cpp:1217:21: note:   no known conversion for argument 1 from ‘<brace-enclosed initializer list>’ to ‘RandomizeMode’
../Makefile_obj:398: recipe for target 'V3Randomize.o' failed
make[2]: *** [V3Randomize.o] Error 1
make[2]: Leaving directory '/home/dguzel/verilator/src/obj_dbg'
Makefile:84: recipe for target '../bin/verilator_bin_dbg' failed
make[1]: *** [../bin/verilator_bin_dbg] Error 2
make[1]: Leaving directory '/home/dguzel/verilator/src'
Makefile:145: recipe for target 'verilator_exe' failed
make: *** [verilator_exe] Error 2
`

Flow is usual git checkout main, git pull main, then autoconf then ./configure then make and this is where I am getting this error. 

ccache version is 3.4.1
gcc version is 8.4.0

Do you have any idea why this is occurs?

Complete output log file can be found in attachments.
[make_err.log](https://github.com/user-attachments/files/16795915/make_err.log)

Thanks

**Comment by wsnyder** [comment]
For build error see #5403 and comment there if that fixes it.

**Comment by DenizzzGuzell** [comment]
Hi, 

Using latest changes and gcc updates. We succesfully verilated the project. However, when simulating it consumes again to many resources and whole server is collapses. Whenever I collect relateable data with this issue. I am going to open new ticket related with this. Before that, it will be great, if you simulate the project given above github repository ``https://github.com/DenizzzGuzell/eFPGA`` eFPGA20x20 and tell me that whether this project can be simulatable with using verilator-main.cpp given in git repository. If not, can you explain why and what I am doing wrong.

Thanks for the help

**Comment by wsnyder** [comment]
I don't know what you are asking, sorry.  For large models you are not expected to trace the entire design, trace just top level to find large problems, then once know what block is bad, trace again a sub-section.

**Comment by DenizzzGuzell** [comment]
Sorry for misunderstanding, What I mean is that in Makefile I have given `--trace-underscore` flag and made only top level module name start with `_`. Also in verilator_main.cpp file I have been tracing the environment using `dut->trace(m_trace, 2);` . These files (Makefile, verilator_main.cpp) is shared with you before in github repository name `eFPGA` and inside that repository you can find the folder given `eFPGA_20x20/sim`. The makefile and verilator_main.cpp can be found inside there. When I try to run that Makefile `make -j -C obj_dir/ -f V_fpga_top.mk V_fpga_top` command creates too many instances of `cc1plus` as you can see below screenshot: 
 
![image](https://github.com/user-attachments/assets/d3292191-f0e1-4421-9e9f-f82e246345de)

I cannot monitor or count how many of them or what is causing of it because these processes are collapses whole server. That is what I need assistance of you whether I am implementing wrong Makefile options or mistaken verilator_main.cpp statements which end up creates too many request or else it is causing because of the complexity and it is not be fixed with flags or metacomments.

Also you have mentioned metacomment related with tracing can you please share a document or usage related with. I could not find a metacomment related with tracing under Verilator arguments.

Thank you

**Comment by wsnyder** [comment]
`--trace-underscore` means turn on tracing of underscore signals, not "only trace underscores".  So you're tracing everything but top ;)

**Comment by DenizzzGuzell** [comment]
I see I totally misunderstood this flag, and I have found trace metacomment. I will try and let you know. Thanks again.

**Comment by DenizzzGuzell** [comment]
Hi Wilson, 

This issue was related with verilating, and we succesfully verilated project after the [commit]([fa32d10](https://github.com/verilator/verilator/commit/fa32d10f398bac73875468a8425f3ce658cc60df))  but simulation again crashes under the condition which only 7 signal is traced using metacomments. We are collecting error message and related data and I am updating GitHub [repository](https://github.com/DenizzzGuzell/eFPGA) to final version we are working on. After that we are going to create a new issue related with simulation. 
Thank you for the help you have given us related with verilating environment.
